examples/integrations/hf_trainer_api/det_callback.py

```diff
@@ -0,0 +1,243 @@
+        self.current_op = next(self.searcher_ops)
+
+        self._check_searcher_compatibility(args)
+
+    def on_log(
+        self,
+        args: TrainingArguments,
+        state: TrainerState,
+        control: TrainerControl,
+        logs=None,
+        **kwargs,
+    ):
+        if state.is_world_process_zero:
+            metrics, metric_type = self._get_metrics(logs)
+            if metric_type == TRAIN:
+                # Prevents reporting metrics for the same step twice. This happens after
+                # training is completed and average training metrics are reported with
+                # the same step as the in-progress training metrics.
+                if self.last_metrics["train_step"] != state.global_step:
+                    self.core_context.train.report_training_metrics(
+                        steps_completed=state.global_step, metrics=metrics
+                    )
+                    metrics["train_step"] = state.global_step
+
+            elif metric_type == EVAL:
+                # Prevents reporting metrics for the same step twice. This happens when
+                # after-training evaluation is completed, and it is reported with the same
+                # step as the last during-training evaluation.
+                if self.last_metrics["eval_step"] != state.global_step:
+                    self.core_context.train.report_validation_metrics(
+                        steps_completed=state.global_step, metrics=metrics
+                    )
+                    metrics["eval_step"] = state.global_step
+                    self.last_metrics.update(metrics)
+            else:
+                logging.warning(f"Metrics not reported: metric type = {metric_type}.")
+
+            self.last_metrics.update(metrics)
+
+        if self.core_context.preempt.should_preempt():
+            control.should_save = True
+
+    def _get_metrics(self, logs: typing.Dict) -> typing.Tuple[typing.Dict, str]:
+        metrics = logs
+        metric_type = get_metric_type(logs)
+        if self.filter_metrics:
+            metrics = {}
+            for k, v in logs.items():
+                if any(m in k for m in self.filter_metrics) is True:
+                    metrics[k] = v
+
+        return metrics, metric_type
+
+    def on_save(
+        self,
+        args: TrainingArguments,
+        state: TrainerState,
+        control: TrainerControl,
+        **kwargs,
+    ):
+        info = det.get_cluster_info()
+        assert info
+
+        # local_path is where HF Trainer saves model and tokenizer.
+        local_path = os.path.join(args.output_dir, f"checkpoint-{state.global_step}")
+
+        if state.is_world_process_zero:
+            if self.user_checkpoint_metadata is not None:
+                self._on_save_user_data(local_path)
+
+        det_checkpoint_metadata = {
+            "steps_completed": state.global_step,
+            "trial_id": info.trial.trial_id,
+        }
+
+        self.core_context.checkpoint.upload(
```

Comment:
i know it's more natural to use `upload` with the hf trainer since it assumes a fixed `output_dir`.  i'm a little concerned about double copy for large models and shared_fs though.  what would using `store_path/restore_path` require and  is it worth it to use it instead of `upload/download`?

Reply:
Let's see if I understand your question correctly. Trainer saves all files to `output_dir`. When we do `upload` for shared_fs, under the hood, we do:
```
def upload(src, dst):
    copytree(src, os.path.join(self._base_path, dst), dirs_exist_ok=True)
```

Let's assume I switch it to `store_path` which gives me the path under which I should save my files. But Trainer already saved my files locally, which means I would need to copy files from `output_dir` to the path:
```
with self.core_context.checkpoint.store_path(metadata) as (path, storage_id):
        copytree(output_dir, path, dirs_exist_ok=True)
```
The only thing to avoid that copying would be to force Trainer to save the files already under the path (not in the output_dir) via args.output_dir = path. 
I looked through Trainer code and I don't see a clean way of modifying args.output_dir dynamically with new paths via callback, especially that we would need to do in the checkpoint context to make sure it works for cloud storage managers.

Reply:
Maybe add a comment here about this issue, in case we revisit in the future?

Reply:
I don't want to mud the waters in the example. I am working on a document describing closer training loop integration with HF Trainer and will include the above discussion there.